framework/src/base/Assembly.C

```diff
@@ -4026,6 +4043,18 @@ Assembly::modifyFaceWeightsDueToXFEM(const Elem * elem, unsigned int side)
   }
 }
 
+unsigned int
+Assembly::residualVectorTagIndex(const TagID tag_id) const
+{
+  for (unsigned int tag_index = 0; tag_index < _residual_vector_tags.size(); ++tag_index)
+    if (_residual_vector_tags[tag_index]._id == tag_id)
+      return tag_index;
```

Comment:
This function will be called many times. It might be a good idea to let `VectorTag` store a `local id (tag index)`. If we do so, we can remove this function. It is a linear algorithm here.